File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.java
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        Solution sol = new Solution();
        sol.solution();
    }
}

class Solution {
	
	int N, answer = 0;
	int[][] map;
	int[][] dirs = {{-1,0},{0,1},{1,0},{0,-1}};

    public void solution() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		map = new int[N][N];

		for(int i = 0; i < N; i ++){
			StringTokenizer st = new StringTokenizer(br.readLine(), " ");
			for(int j = 0; j < N; j ++){
				map[i][j] = Integer.parseInt(st.nextToken());
				answer = Math.max(answer, map[i][j]);
			}
		}
		dfs(0, map);
		System.out.println(answer);
	}

	public void down(int[][] tmp_map) {
		for (int c = 0; c < N; c++) {
			int row = N - 1;
			for (int i = N - 2; i >= 0; i --) {
				if (tmp_map[i][c] > 0) {
					if (tmp_map[i][c] == tmp_map[row][c]) {
						tmp_map[row][c] *= 2;
						answer = Math.max(answer, tmp_map[row][c]);
						tmp_map[i][c] = 0;
						row --;
					}else if(tmp_map[row][c] == 0){
						tmp_map[row][c] = tmp_map[i][c];
						tmp_map[i][c] = 0;
					}else{
						int tmp = tmp_map[i][c];
						tmp_map[i][c] = 0;
						tmp_map[row - 1][c] = tmp;
						row --;
					}
				}
			}
		}
	}

	public void right(int[][] tmp_map){
		for(int i = 0; i < N; i ++){
			int col = N - 1;
			for(int c = N - 2; c >= 0; c --){
				if(tmp_map[i][c] > 0){
					if(tmp_map[i][c] == tmp_map[i][col]){
						tmp_map[i][col] *= 2;
						answer = Math.max(answer, tmp_map[i][col]);
						tmp_map[i][c] = 0;
						col --;
					}else if(tmp_map[i][col] == 0){
						tmp_map[i][col] = tmp_map[i][c];
						tmp_map[i][c] = 0;
					}else{
						int tmp = tmp_map[i][c];
						tmp_map[i][c] = 0;
						tmp_map[i][col - 1] = tmp;
						col --;
					}
				}
			}
		}
	}

	public void up(int[][] tmp_map) {
		for (int c = 0; c < N; c++) {
			int row = 0;
			for (int i = 1; i < N; i++) {
				if (tmp_map[i][c] > 0) {
					if (tmp_map[i][c] == tmp_map[row][c]) {
						tmp_map[row][c] *= 2;
						answer = Math.max(answer, tmp_map[row][c]);
						tmp_map[i][c] = 0;
						row++;
					}else if(tmp_map[row][c] == 0){
						tmp_map[row][c] = tmp_map[i][c];
						tmp_map[i][c] = 0;
					}else{
						int tmp = tmp_map[i][c];
						tmp_map[i][c] = 0;
						tmp_map[row + 1][c] = tmp;
						row ++;
					}
				}
			}
		}
	}

	public void left(int[][] tmp_map){
		for(int i = 0; i < N; i ++){
			int col = 0;
			for(int c = 1; c < N; c ++){
				if(tmp_map[i][c] > 0){
					if(tmp_map[i][c] == tmp_map[i][col]){
						tmp_map[i][col] *= 2;
						answer = Math.max(answer, tmp_map[i][col]);
						tmp_map[i][c] = 0;
						col ++;
					}else if(tmp_map[i][col] == 0){
						tmp_map[i][col] = tmp_map[i][c];
						tmp_map[i][c] = 0;
					}else{
						int tmp = tmp_map[i][c];
						tmp_map[i][c] = 0;
						tmp_map[i][col + 1] = tmp;
						col ++;
					}
				}
			}
		}
	}


	public void dfs(int depth, int[][] cur){
		if(depth == 5){
			return ;
		}

		for(int d = 0; d < 4; d ++){
			int[][] tmp = copy(cur);
			switch(d){
				case 0:
					up(tmp);
					break;
				case 1:
					right(tmp);
					break;
				case 2:
					left(tmp);
					break;
				case 3:
					down(tmp);
					break;
			}
			dfs(depth + 1, tmp);
		}
	}

	public int[][] copy(int[][] m){
		int[][] tmp = new int[N][N];
		for(int r = 0; r < N; r ++){
			for(int c = 0; c < N; c ++){
				tmp[r][c] = m[r][c];
			}
		}

		return tmp;
	}
}

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
import sys
input = sys.stdin.readline

N = int(input())

answer = 0
grid = []
for i in range(N):
    lst = list(map(int, input().split()))
    for j in range(N):
        answer = max(answer, lst[j])
    grid.append(lst)


def left(grid):
    global N, answer
    for row in range(N):
        c = 0
        while c+1 < N:
            if grid[row][c] != 0:
                next_c = c+1
                while c+1 < N and next_c < N:
                    if grid[row][c] == grid[row][next_c]:
                        grid[row][next_c] = 0
                        grid[row][c] *= 2
                        answer = max(answer, grid[row][c])
                        break
                    elif grid[row][next_c] == 0:
                        next_c += 1
                    else:
                        break
            c+=1

    for row in range(N):
        cnt = 0
        for i in range(N):
            if grid[row][i] == 0:
                cnt += 1
            else:
                tmp = grid[row][i]
                grid[row][i] = 0
                grid[row][i-cnt] = tmp

def right(grid):
    global N,answer
    for row in range(N):
        c = N - 1
        while c-1 >= 0:
            if grid[row][c] != 0:
                next_c = c-1
                while c-1 >= 0 and next_c >= 0:
                    if grid[row][c] == grid[row][next_c]:
                        grid[row][next_c] = 0
                        grid[row][c] *= 2
                        answer = max(answer, grid[row][c])
                        break
                    elif grid[row][next_c] == 0:
                        next_c -= 1
                    else:
                        break
            c-=1

    for row in range(N):
        cnt = 0
        for i in range(N-1, -1, -1):
            if grid[row][i] == 0:
                cnt += 1
            else:
                tmp = grid[row][i]
                grid[row][i] = 0
                grid[row][i+cnt] = tmp

def down(grid):
    global N, answer
    for col in range(N):
        r = N - 1
        while r-1 >= 0:
            if grid[r][col] != 0:
                next_r = r-1
                while r-1 >= 0 and next_r >= 0:
                    if grid[r][col] == grid[next_r][col]:
                        grid[next_r][col] = 0
                        grid[r][col] *= 2
                        answer = max(answer, grid[r][col])
                        break
                    elif grid[next_r][col] == 0:
                        next_r -= 1
                    else:
                        break
            r-=1

    for col in range(N):
        cnt = 0
        for i in range(N-1, -1, -1):
            if grid[i][col] == 0:
                cnt += 1
            else:
                tmp = grid[i][col]
                grid[i][col] = 0
                grid[i+cnt][col] = tmp

def up(grid):
    global N, answer
    for col in range(N):
        r = 0
        while r+1 < N:
            if grid[r][col] != 0:
                next_r = r+1
                while r+1 < N and next_r < N:
                    if grid[r][col] == grid[next_r][col]:
                        grid[next_r][col] = 0
                        grid[r][col] *= 2
                        answer = max(answer, grid[r][col])
                        break
                    elif grid[next_r][col] == 0:
                        next_r += 1
                    else:
                        break
            r+=1

    for col in range(N):
        cnt = 0
        for i in range(N):
            if grid[i][col] == 0:
                cnt += 1
            else:
                tmp = grid[i][col]
                grid[i][col] = 0
                grid[i-cnt][col] = tmp

def change_grid(d, grid):
    if d == 0:
        up(grid)
    elif d == 1:
        right(grid)
    elif d == 2:
        down(grid)
    else:
        left(grid)

def copy(grid):
    tmp = [[0] * N for _ in range(N)]
    for r in range(N):
        for c in range(N):
            tmp[r][c] = grid[r][c]

    return tmp

def dfs(depth, grid):
    if depth == 5:
        return
    
    for d in range(4):
        tmp = copy(grid)
        change_grid(d, tmp)
        dfs(depth + 1, tmp)
        
dfs(0, grid)
print(answer)